server/src/sql-tools/schema-from-decorators.ts

```diff
@@ -0,0 +1,444 @@
+const asCheckConstraintName = (table: string, expression: string) => asKey('CHK_', table, [expression]);
+const asIndexName = (table: string, columns?: string[], where?: string) => {
+  const items: string[] = [];
+  for (const columnName of columns ?? []) {
+    items.push(columnName);
+  }
+
+  if (where) {
+    items.push(where);
+  }
+
+  return asKey('IDX_', table, items);
+};
+
+const makeColumn = ({
+  name,
+  tableName,
+  options,
+}: {
+  name: string;
+  tableName: string;
+  options: ColumnOptions;
+}): DatabaseColumn => {
+  const columnName = options.name ?? name;
+  const enumName = options.enumName ?? `${tableName}_${columnName}_enum`.toLowerCase();
+  let defaultValue = asDefaultValue(options);
+  let nullable = options.nullable ?? false;
+
+  if (defaultValue === null) {
+    nullable = true;
+    defaultValue = undefined;
+  }
+
+  const isEnum = !!options.enum;
+
+  return {
+    name: columnName,
+    tableName,
+    primary: options.primary ?? false,
+    default: defaultValue,
+    nullable,
+    enumName: isEnum ? enumName : undefined,
+    enumValues: isEnum ? Object.values(options.enum as object) : undefined,
+    isArray: options.array ?? false,
+    type: isEnum ? 'enum' : options.type || 'character varying',
+    synchronize: options.synchronize ?? true,
+  };
+};
+
+const asDefaultValue = (options: { nullable?: boolean; default?: ColumnDefaultValue }) => {
+  if (typeof options.default === 'function') {
+    return options.default() as string;
+  }
+
+  if (options.default === undefined) {
+    return;
+  }
+
+  const value = options.default;
+
+  if (value === null) {
+    return value;
+  }
+
+  if (typeof value === 'number') {
+    return String(value);
+  }
+
+  if (typeof value === 'boolean') {
+    return value ? 'true' : 'false';
+  }
+
+  if (value instanceof Date) {
+    return `'${value.toISOString()}'`;
+  }
+
+  return `'${String(value)}'`;
+};
+
+const missingTableError = (context: string, object: object, propertyName?: string | symbol) => {
+  const label = object.constructor.name + (propertyName ? '.' + String(propertyName) : '');
+  return `[${context}] Unable to find table (${label})`;
+};
+
+// match TypeORM
+const sha1 = (value: string) => createHash('sha1').update(value).digest('hex');
+
+const findByName = <T extends { name: string }>(items: T[], name?: string) =>
+  name ? items.find((item) => item.name === name) : undefined;
+const resolveTable = (tables: SchemaTables, object: object) =>
+  findByName(tables, Reflect.getMetadata(SchemaKey.TableName, object));
+
+let initialized = false;
+let schema: DatabaseSchema;
+
+export const reset = () => {
+  initialized = false;
+  items.length = 0;
+};
+
+export const schemaFromDecorators = () => {
+  if (!initialized) {
+    const schemaTables: SchemaTables = [];
+
+    const warnings: string[] = [];
+    const warn = (message: string) => void warnings.push(message);
+
+    for (const { item } of items.filter((item) => item.type === 'table')) {
```

Comment:
This is pretty inefficient since you filter the same items over and over again. What do you think of iterating over items first and adding it to an `(item type, [items])` map?

Reply:
I think it should be possible to keep the types sane but given we can easily assume <100 columns per table you're right and that really doesn't matter.